### DISCUSSION TITLE
Add support for a C library exposing the filesystem API

### DISCUSSION THREAD

**Discussion by cipriancraciun** [mhx/dwarfs#22] [Ideas]
I've taken a very quick look at the `dwarfs` FUSE application, and from what I've seen the code is extremely simple and clean, and basically just delegates operations to a `filesystem_v2` object.

So here is a thought:  how about writing a small C (not C++) wrapper for the `filesystem_v2` class.  Basically instead of `fs.find(some_path)` one would write `dwarfs_fs2_find(fs, some_path)`;  granted one now needs to provide similar accessor functions for `entry_view` (and derived).

Basically the C library should cover only the file-system "read", and provide only the most basic operations, similar to what FUSE needs.  (In fact I think one could just hack the `dwarfs.cpp` and just fake FUSE calls to obtain such a library.)

----

What's the point you ask?  :)

Using DwarFS library inside other applications to access files stored in a `dwarfs` image, without mounting it via FUSE.

For example:
* an HTTP server that serves files directly from the image;  BTW, I have implemented something like this in Go, based on CDB, https://github.com/volution/kawipiko, although CDB is limited to 4 GiB and compression was achieved per file;
* an application having lots of "resource" files (think images, sounds, scripts, etc.) could just access them from the image, thus simplifying the deployment;  (i.e. copy one file instead of 10K;)
* an AI / machine-learning trainer that needs large text-based datasets for training;
* as an alternative for a key-value store, but one that provides good compression and deduplication;

Why a C library and not C++?  It's easier to integrate it in other languages like Go, Rust, Erlang, etc.

**Comment by cipriancraciun** [comment]
BTW, GitHub has just opened "Discussions" to all public repositories (it must be enabled from the repository settings).  Thus for example this "issue" would have been better discussed in there, as it's not actually a "bug" or clear "feature request".

**Comment by mhx** [comment]
At least the C++ library already exists: `libdwarfs.a`. The various binaries are really just relatively thin wrappers around the library. Both the read and write parts are exposed, you can even build a dwarfs image in-memory without having to access any real files.

Indeed, a C API could probably done quite easily (at least for the read part).
